src/useSubscription.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  ApolloClient,
+  ApolloError,
+  OperationVariables,
+  SubscriptionOptions,
+} from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { useEffect, useRef, useState } from 'react';
+import { useApolloClient } from './ApolloContext';
+import { Omit, objToKey } from './utils';
+
+export type OnSubscriptionData<TData> = (
+  options: OnSubscriptionDataOptions<TData>
+) => any;
+
+export interface OnSubscriptionDataOptions<TData> {
+  client: ApolloClient<any>;
+  subscriptionData: SubscriptionHookResult<TData>;
+}
+
+export interface SubscriptionHookOptions<TData, TVariables>
+  extends Omit<SubscriptionOptions<TVariables>, 'query'> {
+  onSubscriptionData?: OnSubscriptionData<TData>;
+}
+
+export interface SubscriptionHookResult<TData> {
+  data?: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export function useSubscription<TData = any, TVariables = OperationVariables>(
+  query: DocumentNode,
+  {
+    variables,
+    onSubscriptionData,
+    ...options
+  }: SubscriptionHookOptions<TData, TVariables> = {}
+): SubscriptionHookResult<TData> {
+  const client = useApolloClient();
+  const onSubscriptionDataRef = useRef<null | OnSubscriptionData<TData>>(null);
+  const [result, setResult] = useState<SubscriptionHookResult<TData>>({
+    loading: true,
+  });
+
+  if (onSubscriptionData) {
+    onSubscriptionDataRef.current = onSubscriptionData;
+  }
+
+  useEffect(
+    () => {
+      const subscription = client
+        .subscribe({
+          ...options,
+          query,
+          variables,
+        })
+        .subscribe({
+          error: error => {
+            setResult({ loading: false, data: result.data, error });
+          },
+          next: nextResult => {
+            const newResult = {
+              data: nextResult.data,
+              error: undefined,
+              loading: false,
+            };
+            setResult(newResult);
```

Comment:
Using `setResult` here causes React warnings. Until https://github.com/facebook/react/issues/14769 is fixed I recommend a workaround similar to the one in `useQuery`: https://github.com/trojanowski/react-apollo-hooks/blob/6c9c90fa5fcb349b3e3e308bfb93901e59af6a3a/src/useQuery.ts#L157

In this file it could be applied that way:

1. I'd rename the function for setting state returned  by `useState` to e. g. `setResultBase`.
2. I'd define `setResult` like this:

```
  function setResult(newResult: SubscriptionHookResult<TData>) {
    actHack(() => {
      setResultBase(newResult);
    });
  }
```

`actHack` is available with `import actHack from './internal/actHack';`.

After https://github.com/facebook/react/pull/14853 we should be able to get rid of those hacks.